### PR TITLE
Rebuild the Site

### LINKED ISSUE #2018: Build time is insufferably long
This is my laptop building the documentation site:

image

This total resource consumption has been going on for around 20 minutes now.

Further more I see the build times tripling in travis:

image

It looks like https://github.com/webpack/webpack.js.org/commit/b45b73e1a4f3d600f006fec5cc95de20aa9b7382 introduced a major performance regression

### PR DESCRIPTION
**Pull request by skipjack**
This PR is a complete refactor of the site's build process using pure `webpack`. It offers...

- Much faster build times (~30s).
- More PWA features (e.g. the site now acts like an SPA once entered).
- Various code simplification and cleanup.

I let this branch get a little too big with not enough visibility. This PR should remedy the visibility part and start us on a path to get it wrapped up in the next few weeks so we can go back to focusing on content. Here's what's left:

- [x] Finish re-incorporating mobile sidebar
- [ ] Update CONTRIBUTING.md file to better explain the new flow (@jeremenichelli)
- [x] Extract anchors into `_content.json` via `DirectoryTreePlugin` (@montogeek - #2041)
- [x] Finish porting the markdown process to `remark`
- [x] Re-integrate google-analytics (@montogeek - #2045)
- [x] Populate html files with content instead of loading
- [x] Fix external fetch process (@jeremenichelli - #2072)
- [x] Re-incorporate `redirects.json` (@EugeneHlushko - #2034)
- [x] Populate page title in `server.jsx` (@montogeek - #2040)
- [x] Finish `Navigation` component (@skipjack - #2048)
- [x] Add custom route for landing page (@skipjack)

Nice to haves (won't be blockers for release and can be tackled after merge):
- [ ] Preload bundles for future navigation
- [ ] Improve path array scrapping for page title population
- [x] Gather data metrics of current site on different sections for later comparison with rebuilt version (@jeremenichelli)
- [ ] Improve font loading strategy (@jeremenichelli)
- [ ] Replace font icons with SVGs (@jeremenichelli)

No more work should be done on this branch directly. Please open PRs directed towards `rebuild` to hit these TODOs.

This work is heavily related to #1525 and #1400. Once this is merged we should probably close both those tickets and make more targeted tickets/prs for anything else that remains. I think this also closes #2018 and probably a few other tickets. With the introduction of `remark-mermaid`, it makes #469 finishable. Please comment if you know of other tickets this might resolve.

### CODE REVIEW

**Comment by skipjack** [comment]
@montogeek I pushed your commits to branch called `rebuild-montogeek`. If you could split that into individual PRs against this branch and assign @jeremenichelli, @EugeneHlushko and I as reviewers that would be great.

@EugeneHlushko maybe rebase #2034 on the latest `rebuild` and assign us as reviewers.

I'm going to try to wrap up navigation work today... I did some refactoring to `react-banner` which should make this much easier.

**Comment by EugeneHlushko** [comment]
@skipjack its already rebased on current ver 👍

**Comment by montogeek** [comment]
You added `package-lock.json`, Want to replace yarn?

**Comment by montogeek** [comment]
@skipjack I actually had this conversation with @ooflorent regarding how webpack repos use npm and yarn, core repo uses yarn, new repos like mini-extract-css uses yarn, but some others use npm.

It would be nice to use only one, will make things easier for new contributors which sometimes push package-lock and yarn.lock files in PRs.

Personally I am fine with either, just speaking for the community. In my personal projects I always pin dependencies versions.

**Comment by EugeneHlushko** [comment]
i would vote for `npm` as it got pretty good after landing v5 and .3+ patches. But as @montogeek noted other repost under webpack use yarn, and maintaining 2 lockfiles isn't fun at all i would suggest just using `yarn`

File: package.json [hunk 1]
```diff
@@ -24,48 +24,47 @@
...
-    "start": "npm run init:generated && npm run fetch && node ./antwar.bootstrap.js develop",
-    "start-only": "node ./antwar.bootstrap.js develop",
-    "build": "npm run init:generated && npm run fetch && rm -rf build/ && node ./antwar.bootstrap.js build && npm run sitemap && echo webpack.js.org > build/CNAME",
-    "build-test": "npm run build && http-server build/",
-    "deploy": "gh-pages -d build",
+    "start": "webpack-dev-server --config webpack.dev.js --env.dev",
+    "build": "webpack --config webpack.prod.js",
     "fetch": "sh src/scripts/fetch.sh",
-    "init:generated": "mkdirp ./generated/loaders && mkdirp ./generated/plugins ",
+    "test": "npm run lint",
```
Comment (montogeek) [inline package.json]: I would like to have e2e tests, using puppeteer or cypress.

File: package.json [hunk 2]
```diff
@@ -84,13 +84,18 @@
...
     "sass-loader": "^6.0.6",
     "sitemap-static": "^0.4.2",
+    "static-site-generator-webpack-plugin": "^3.4.1",
     "style-loader": "^0.18.2",
     "tap-min": "^1.2.1",
     "tap-parser": "^6.0.1",
     "through2": "^2.0.3",
+    "uglifyjs-webpack-plugin": "^1.1.6",
     "url-loader": "^0.5.9",
     "webpack": "^3.10.0",
```
Comment (dhruvdutt) [inline package.json]: Can we also update to use webpack 4? :grin:

Comment (EugeneHlushko) [reply package.json]: @dhruvdutt yeah but it was decided to do as a follow up right after we merge this into master to not hold it anymore. I will just apply part of my WP4 changes from master into newer `master` after this merged so its on webpack 4 and also update the configs according to new build setup

Comment (dhruvdutt) [reply package.json]: Sounds like a plan. :100:

**Comment by EugeneHlushko** [comment]
@skipjack is `Rethink external population process` a blocker for this?

**Comment by skipjack** [comment]
>  is Rethink external population process a blocker for this?

@EugeneHlushko yes, at least in these ways...

- We don't need any `.json` files anymore (not even sure how they were used previously).
- All external `.md` files should populate the proper location within `/src/content`.
- Each generated file should be prefixed with an `_` in order to be `.gitignore`d.
- We'll need to update `_content.json` generation so underscores are stripped from the `url` but remain in the `path` so they can be found on `import`.

There's probably a variety of other cleanup to be done but I think the steps above are the bare minimum to get this merged. I'd be happy to help out with this although I think @jeremenichelli had some interest in tackling it.

**Comment by jeremenichelli** [comment]
I've noticed that in the _server file_ we are using `require` instead of `import` as in the _client index file_, meaning that the intention is to populate the content on the html file. This is sadly not working, I'm seeing a `[object Object]` expression in the dist files.

I tried to solve it but couldn't. I will give it a try again tomorrow, but just in case someone ones to try to solve it too and has more luck I added the task in the description.

**Comment by montogeek** [comment]
@jeremenichelli @skipjack Already solved it, What did you tried?

**Comment by jeremenichelli** [comment]
@montogeek @skipjack I solved the issue locally, now the other problem is that we are anyway overriding the content even when it was server side rendered. I can tell yoiu that solving this will be super tricky and it already feels we are building our own version of Gatsby 😅 but I'll give it a try, we need this to be solved because the user sees the content and it flashes when the bundle gets executed.

**Comment by skipjack** [comment]
Ok another few things down (fixed big issue with initial load discovered by @jeremenichelli and finished title population). The todos in the description are updated. I'm going to try to review, help finish,  and merge the rest of the PRs by the end of this week. Then we can start talking about getting this branch merged to `master`.

**Comment by EugeneHlushko** [comment]
Any updates here? Anything i could help with?

**Comment by skipjack** [comment]
@EugeneHlushko you could finish the `SidebarMobile` component and aside from that there's only:

- [ ] Figure out whether we want to support links within code blocks (see `-with-links` usage).
- [ ] Figure out whether we want to support details within code blocks (see `-with-details` usage).
- [ ] Decide if we care about the mobile table improvements from `master` in `markdown.js`. It made table formatting on smaller screens nicer but introduced a lot of complexity. Personally, I recommend we leave it out for now and if it's requested again we should just ask that someone writes a `remark` plugin for mobile table styling.

Other than that I would call this merge-able but I'll leave it to @jeremenichelli and you guys (@montogeek and @EugeneHlushko) to make the final call. It would also be nice to get the `hyperlink` improvements @Munter made on #1582 into this branch but that could also happen after this is merged.

I have very little bandwidth at this point to help anymore but like I said I think this in a pretty good place. Feel free to ping me though if there's anything that's still up in the air.

**Comment by jeremenichelli** [comment]
@skipjack I think we are in a really good shape. 

 - About the missing Prism processing types, my call is that we should support the current state of it and parse them somehow, if not we would need to rewrite those sections which I'm not against to, but my be even more work. We should make a decision around it.
 - Other issue, build is failing locally to me because we are trying to read a huge JSON file (`content.json`) so we might need to move it to a stream, I can try to accomplish that tomorrow.
 - Then we should go over all the site and make sure we are not degrading the experience in any device width.
 - And finally, merge master and solve the conflicts there, I can also do that when we've already covered the rest of the tasks.

I will wait for your call on these points @EugeneHlushko and @montogeek.

**Comment by montogeek** [comment]
I merged master into this branch. Site works locally and after building it.

**Comment by montogeek** [comment]
I broke the build when merging master, will fix asap

File: package.json (hunk 1 above)
Comment (jeremenichelli) [reply package.json]: Though definitely out of scope here, sounds like an overkill for linting `*.md` files, in which situation you think e2e tests will cover something we wouldn't be able to with this setup?

File: webpack.config.js [hunk 3]
```diff
@@ -1,135 +0,0 @@
-const path = require('path');
```
Comment (jeremenichelli) [inline webpack.config.js]: Was this file unused?

Comment (EugeneHlushko) [reply webpack.config.js]: seems now we only use webpack.common.js webpack.prod.js and webpack.dev.js

Comment (montogeek) [reply webpack.config.js]: @jeremenichelli https://github.com/webpack/webpack.js.org/pull/2251

Comment (jeremenichelli) [reply webpack.config.js]: 👍

**Comment by Legends** [comment]
@skipjack @montogeek 
Is it currently possible to run `rebuild` locally?

Have to remove `NODE_ENV=development` here:
> "start": "NODE_ENV=development webpack-dev-server --config webpack.dev.js --env.dev",

`supporters.json` is missing
>  Can't resolve './_supporters.json' in 'C:\Users\xx\Documents\_Github_Repositories\webpack.js.org\src\components\Support'

..probably more things exist preventing build.

**Comment by jeremenichelli** [comment]
@Legends make sure you have a clean repository start, run `yarn` to install branch dependencies, run `yarn fetch` and then `yarn start`. With no modifications on the commands it should work right out of the box.

**Comment by montogeek** [comment]
@Legends You are using Linux, that `NODE_ENV` is common in unix systems to declare a environment variable for the command. We would probably need a package to handle that.

To run the site, make sure to delete node_modules, install then again, then run `fetch` and `content` script, after that you can run `start`.

**Comment by Legends** [comment]
@montogeek 
I guess, to check for windows it's like: `const isWindows = process.platform==='win32'`.

Under windows it would be:
>  "start": "set NODE_ENV=development && webpack-dev-server --config webpack.dev.js --env.dev"

<hr/>
Webpack compiles successfully now, don't know if it's because of the executed `content` command or the deleted node_modules folder. 

When I open [http://localhost:3000/](http://localhost:3000/) I get:

image

**in source:**
image

**Comment by byzyk** [comment]
@montogeek @Legends you might wanna have a look at this: [`cross-env`](https://github.com/kentcdodds/cross-env#readme). It handles the mismatch between Win and Unix platforms.

**Comment by skipjack** [comment]
@wooorm any recommendations for how we can write `remark` plugins to support code block extensions like the ones mentioned above:

> - Figure out whether we want to support links within code blocks (see `-with-links` usage).
> - Figure out whether we want to support details within code blocks (see `-with-details` usage).

@montogeek I know you mentioned maybe taking another route, but I think if there's a clean way to do it through remark we should try it. I definitely understand how this would be a bit wonky to implement I figured but @wooorm might have some ideas.

**Comment by wooorm** [comment]
@skipjack I’d like tot help, but it’s a big thread, not sure which extensions you’re talking about?

**Comment by skipjack** [comment]
Yeah, sorry let me try to explain...

A contributor (I think maybe even @sokra) customized our `marked` parsing to allow links (i.e. `[link](..)`) and `<details><summary>...</summary>......</details>` usage within code blocks. The way this is currently enabled in a code block is by adding to the language string, i.e.

````` md
``` js-with-links
{
  [linked-property](...): true
}
```

``` js-with-details
{
  <details><summary>
  'collapsed-property': {
  </summary>
     some: false,
     hidden: false,
     fields: true
  }
}
```
`````

**Comment by montogeek** [comment]
@skipjack I already did it the first one (#2338 ) and contacted @wooorm about `details` one.
I actually would like to change how `-with-details` works, maybe moving that _markdown_ to another format.

**Comment by montogeek** [comment]
That is actually the last thing left to implement before merging this branch. It is the only blocker. Everything else works.

**Comment by skipjack** [comment]
@montogeek ah nice, sorry I didn't realize you ended up making progress on this. Feel free to ignore then.

**Comment by montogeek** [comment]
@skipjack Not worries, but since it is my first time using `unified` ecosystem, I am pretty sure @wooorm could have some better ideas about my plugins :D

**Comment by Legends** [comment]
@montogeek , @jeremenichelli 
After:
1. deleting `node_modules`
2. `yarn install`
3. `yarn fetch`
4. `yarn content`
5. `yarn start`

But I get the error I mentioned [above](https://github.com/webpack/webpack.js.org/pull/2039#issuecomment-406821825).
Do I have to configure/change something else?

**Comment by montogeek** [comment]
@Legends Could you please create a separate issue to discuss it, this PR is already big enough and I don't want to add more noise.

File: package.json (hunk 1 above)
Comment (montogeek) [reply package.json]: Just to verify that the page works, all pages are there, Gitter chat works, mobile sidebar works, Search too, etc.

**Comment by montogeek** [comment]
@Munter We broke the website: https://travis-ci.org/webpack/webpack.js.org/jobs/411722550

**Comment by montogeek** [comment]
I am going to deploy latest master, Are you available to fix the script?

File: .travis.yml [hunk 4]
```diff
@@ -1,15 +1,52 @@
...
+      cache:
+        pip: true
+        directories:
+          - $HOME/.cache
+      install: pip install -r requirements.txt
+      script: cp .proselintrc ~/ && proselint src/content
+
+    - stage: Post-build
+      name: Deploy
+      if: branch = master
```
Comment (montogeek) [inline .travis.yml]: @Munter This isn't really working, no idea why:
https://docs.travis-ci.com/user/conditional-builds-stages-jobs/ says:
> (note for PRs this is the base branch name)

but this PR base branch is `master`. :/

**Comment by montogeek** [comment]
But in the bright side, it works really well :D

File: .travis.yml (hunk 4 above)
Comment (Munter) [reply .travis.yml]: I have no clue why it deployed there. It's literally copy pasted from the example that shows how to only deploy when building the master branch.

**Comment by Munter** [comment]
I don't have a laptop with me on vacation, so the earliest I can look at this is Monday. But I don't know how to fix it, since I'm just following the Travis docs on that conditional check. I have no idea why it wouldn't filter out that job on a non-master branch

**Comment by Munter** [comment]
This might work:

’if: branch = master AND type = push’

I'm on a mobile phone, so I'm not going to attempt a fix myself

**Comment by montogeek** [comment]
@Munter Sorry for disturbing your vacations, I will try later.

**Comment by Munter** [comment]
I think this has been the case all the time. One for checking the branch push and one for testing the hypothetical merge into the base branch

**Comment by montogeek** [comment]
@Munter No, it started creating 2 jobs since your travis PR

**Comment by Munter** [comment]
The Travis settings are responsible for that:

image

**Comment by montogeek** [comment]
This PR https://github.com/webpack/webpack.js.org/pull/2422 was from the repo but it didn't create 2 jobs.
I just want to understand how Travis works. Feel free to ping me on Slack, sorry for the noise. :D

**Comment by Munter** [comment]
These travis settings on #2422 might be the reason:

```yml
branches:
  only:
    - master
    - develop
```

**Comment by Munter** [comment]
Should we evaluate the status of this PR?

I think the mobile sidebar point can be checked off the list. Which basically leaves documentation. All the remaining points have been marked as optional, right?

What's left to do to get his merged?

**Comment by Munter** [comment]
This branch seems to have stalled. It's a shame if all the work everyone has put into it gets outdated. The feedback loop for contributors is still really slow, which this branch fixes.

Are the changes that are still needed really blocking a merge?

**Comment by EugeneHlushko** [comment]
Yes there is one outstanding issue, @montogeek had opened a WIP PR

**Comment by montogeek** [comment]
Will work on this this weekend.
There is more things I would like to include, but finishing last 2 blockers is fine.

**Comment by Munter** [comment]
I think I can find some time between Christmas and new years to pitch in as well, so we can get the lats bits in

**Comment by montogeek** [comment]
Update regarding https://github.com/webpack/webpack.js.org/pull/2039#issuecomment-425888685
Once https://github.com/webpack/webpack.js.org/pull/2564 is merged, we can merge this branch.

**Comment by EugeneHlushko** [comment]
Need to look why travis failing tho

**Comment by EugeneHlushko** [comment]
The links linter is still failing, and if i understand correctly:
- you had opened a PR to fix external MD file that fails to lint
- next time external .md file gets changes, there is change our build stops working

Please correct me if i am wrong @Munter

**Comment by Munter** [comment]
@EugeneHlushko Consistently problematic external dependencies, like badges, which have shown instability, can be marked as `--skip` or `--todo` in hyperlink, which allows us to either not to the test at all for a url, or accept an error on a url.

The problem here was, there was no url at all, see https://github.com/webpack-contrib/cache-loader/pull/60

That triggered what I would call a bug in html-loader, which means there was no real URL to match against for a `--skip` or `--todo` pattern: https://github.com/webpack-contrib/html-loader/issues/191

**Comment by EugeneHlushko** [comment]
Maybe this is not the right way to go about it? Could be that doing checks on the links that this repo owns in ./src/**/*.md before fetching external content?

**Comment by EugeneHlushko** [comment]
I think that our cd is at risk with this tho as i pointed out earlier, if the rest of maintainers are happy to deal with that, lets leave it as you implemented - to check externally fetched md files too

**Comment by EugeneHlushko** [comment]
Let's ping maintainers again for final look since no one responded and merge in X days if no additional feedback is given. This is a huge change that everyone had worked hard on and waited for.

cc @jeremenichelli @skipjack @montogeek @Munter

**Comment by skipjack** [comment]
Awesome to see this finally moving forward again, thanks to those that revived it. I’d love to help more but am still very strapped for time.

If I remember correctly, it was only one or two markdown parsing issues that were blocking the merge. Based on the other comments/merges, it sounds like those are resolved so from my standpoint this is good to go. If we notice anything after the merge I’ll try to find some time to help out.

**Comment by montogeek** [comment]
@skipjack It is always nice hearing from you. Yes, it is ready but we haven't found or schedule the time to do it.
We need to find some time, maybe this Saturday?

**Comment by Munter** [comment]
I should be able to help out Saturday in case things explode

**Comment by EugeneHlushko** [comment]
Lets see how it goes... 🤞

**Comment by jeremenichelli** [comment]
OMG IT'S HAPPENING

**Comment by EugeneHlushko** [comment]
Its live!

**Comment by jeremenichelli** [comment]
Kudos on everyone involved, it's working like a charm <3

**Comment by skipjack** [comment]
Yeah just poked around a bit, seems stable. Tested navigating through a variety of sections and searching, all seemed to work fine. The only thing I'm noticing, which is definitely not critical, is that the sidebar only sticks for a little bit and then floats up with the rest of the page on scroll. Probably a little css tweak required there.

I checked the console and the only errors I'm seeing are `images.opencollective` 404s on the landing page (which I think are to be expected).

**Comment by montogeek** [comment]
@EugeneHlushko Nice!

**Comment by Munter** [comment]
Awesome!!!

Since we also eliminated the continuous github API polling on this branch, we can now turn off the Travis parallel run limitation that we had previously imposed. That should cut roundtrip time for feedback on PR builds down by quite a lot combined with the already shorter build times

**Pull request merged**